### DISCUSSION TITLE
Best practices for migrating the database from client running in Kubernetes

### DISCUSSION THREAD

**Discussion by trenslow** [ravendb/ravendb#14415] [Q&A] [answered]
Hi community,

We currently have a problem migrating the schema of a RavenDB database via a .NET service running in Kubernetes using the [RavenMigrations](https://github.com/migrating-ravens/RavenMigrations) package. While the migrations are running, the service doesn't respond to the Kubernetes liveness check. Kubelet then kills the container, causing the migrations to stop mid-way and put the database in a broken state. After some research a couple ideas came up:

1) We can increase the time that it takes for Kubernetes to check the liveness of the container. However, that's not ideal, as we don't know ahead of time how large the database is and therefore how long the migrations will take, not to mention that the size of the worker node that the service runs on could potentially vary, leading to increased processing time.

2) We can run the migrations in an init container. While there's generally nothing speaking against it, it would be a considerable time investment to get it up and running.

Does anyone have experience with running RavenDB migrations from a client running in Kubernetes who could advise us on the most robust solution? Is there a better way to migrate the schema that we haven't considered? 

Thanks for any advice

**Comment by gregolsky** [comment]
Do you know what does that Kubernetes liveness check do exactly?
I see there's a ticket to [make RavenMigrations async](https://github.com/migrating-ravens/RavenMigrations/issues/51), maybe that's the issue it does not respond to that check?

Have you tried asking on [RavenMigrations github](https://github.com/migrating-ravens/RavenMigrations)?

**Reply by trenslow** [reply]
The liveness check lets Kubelet know that a container started properly and also hasn't reached some sort of deadlock. The readiness check is so that Kubelet knows whether to send traffic to a pod/container. In the case of the migrations, we would want the liveness check to respond with 200 OK and the readiness check to give a non 200 OK response while migrations are running. Only when they finish running successfully should both endpoints respond with 200 OK.

The issue is that the migrations run before the server even starts up, causing neither endpoint to respond. Could be that it has to do with not being async, but there still should be some way to do it synchronously I guess.

I planned to ask the RavenMigrations guys but I wanted to ask here first what the RavenDB maintainers recommend for migrations and what sort of experience they have with them.

**Comment by ayende** [comment, marked as answer]
Your problem is likely that you are running the migration on startup? Blocking the actual application start, no?
K8s then ping the node, see it isn't ready, and kill it.

Can you do something like this:

`Program.cs`
```
var migrateTask = Task.Run(() => {
        // do migration
});

[Action("/k8s/health")]
Task<int> HealthCheck()
{
    return 200;
}

[Action("/k8s/ready")]
Task<int> Ready()
{
    return migrateTask.IsCompleted ? 200 : 429;
}

```

**Reply by trenslow** [reply]
You're exactly right @ayende. We will give it a shot and let you know.
